warehouse/legacy/api/pypi.py

```diff
@@ -731,21 +731,22 @@ def file_upload(request):
         # go along.
         with open(temporary_filename, "wb") as fp:
             file_size = 0
-            file_hash = hashlib.md5()
+            file_hashes = {"md5": hashlib.md5(), "sha256": hashlib.sha256()}
             for chunk in iter(
                     lambda: request.POST["content"].file.read(8096), b""):
                 file_size += len(chunk)
                 if file_size > file_size_limit:
                     raise _exc_with_message(HTTPBadRequest, "File too large.")
                 fp.write(chunk)
-                file_hash.update(chunk)
+                for hasher in file_hashes.values():
+                    hasher.update(chunk)
 
         # Actually verify that the md5 hash of the file matches the expected
         # md5 hash. We probably don't actually need to use hmac.compare_digest
         # here since both the md5_digest and the file whose file_hash we've
         # computed comes from the remote user, however better safe than sorry.
         if not hmac.compare_digest(
-                form.md5_digest.data, file_hash.hexdigest()):
+                form.md5_digest.data, file_hashes["md5"].hexdigest()):
```

Comment:
Although I was going to require it to match *both* of them (if both are given) and require at least one of them.